File: exercises/quarkus/shipping-service/src/main/java/com/example/shipping/application/ScheduleShipmentUseCase.java
package com.example.shipping.application;

import com.example.shipping.domain.identifier.ShippingContextKey;
import com.example.shipping.domain.model.Shipment;
import com.example.workshop.observability.BaggageHelpers;
import com.example.workshop.observability.DomainContext;

import io.micrometer.core.instrument.MeterRegistry;
import io.opentelemetry.api.trace.Span;
import io.opentelemetry.instrumentation.annotations.WithSpan;
import jakarta.enterprise.context.ApplicationScoped;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

@ApplicationScoped
public class ScheduleShipmentUseCase {

    private static final Logger log = LoggerFactory.getLogger(ScheduleShipmentUseCase.class);

    private final MeterRegistry meterRegistry;

    public ScheduleShipmentUseCase(MeterRegistry meterRegistry) {
        this.meterRegistry = meterRegistry;
    }

    @WithSpan("Shipping.Schedule")
    public Shipment schedule(ScheduleShipmentCommand command) {

        try (var ctx = DomainContext.open(
                ShippingContextKey.ORDER_ID.of(command.orderId()),
                ShippingContextKey.CUSTOMER_ID.of(command.customerId()))) {

            String tier = BaggageHelpers.get("customer.tier");
            if (tier == null) tier = "unknown";

            Span span = Span.current();
            span.setAttribute("order.id", command.orderId());
            span.setAttribute("customer.id", command.customerId());
            span.setAttribute("customer.tier", tier);
            span.setAttribute("shipping.class", command.shippingClass());
            span.setAttribute("shipping.line_items_count", command.totalLineItemCount());

            Shipment shipment = Shipment.schedule(
                    command.orderId(),
                    command.shippingClass(),
                    command.totalLineItemCount());

            ctx.include(ShippingContextKey.SHIPMENT_ID.of(shipment.id().value()));
            span.setAttribute("shipment.id", shipment.id().value());

            log.info("Shipment scheduled: {} (class={} items={})",
                    shipment.id(), command.shippingClass(), command.totalLineItemCount());

            meterRegistry.counter("shipping_shipments_scheduled_total",
                    "class", command.shippingClass(),
                    "tier", tier).increment();

            return shipment;
        }
    }
}

File: exercises/dotnet/ShippingService/Application/ScheduleShipmentUseCase.cs
using System.Diagnostics;
using System.Diagnostics.Metrics;
using ShippingService.Domain.Identifiers;
using ShippingService.Domain.Models;
using SharedObservability;

namespace ShippingService.Application;

/// <summary>
/// The schedule-shipment use case. Schedules a shipment for a fulfilled order.
/// Shipping always succeeds in this workshop model -- the outcome is always SCHEDULED.
/// </summary>
public class ScheduleShipmentUseCase
{
    private static readonly ActivitySource Source = new("ShippingService");
    private static readonly Meter ShippingMeter = new("ShippingService");
    private static readonly Counter<long> ShipmentsCounter =
        ShippingMeter.CreateCounter<long>("shipping_shipments_scheduled_total",
            description: "Total shipments scheduled by class and tier");

    private readonly ILogger<ScheduleShipmentUseCase> _logger;

    public ScheduleShipmentUseCase(ILogger<ScheduleShipmentUseCase> logger)
    {
        _logger = logger;
    }

    public Shipment Schedule(ScheduleShipmentCommand command)
    {
        using var activity = Source.StartActivity("Shipping.Schedule");

        using var ctx = new DomainContext(_logger,
            ShippingContextKey.OrderId.Of(command.OrderId),
            ShippingContextKey.CustomerId.Of(command.CustomerId));

        var tier = BaggageHelpers.Get("customer.tier") ?? "unknown";

        activity?.SetTag("order.id", command.OrderId);
        activity?.SetTag("customer.id", command.CustomerId);
        activity?.SetTag("customer.tier", tier);
        activity?.SetTag("shipping.class", command.ShippingClass);

        var shipment = Shipment.Schedule(command.OrderId, command.ShippingClass);

        // Enrich context with result
        BaggageHelpers.Set("shipment.id", shipment.Id.Value);
        activity?.SetTag("shipment.id", shipment.Id.Value);
        activity?.SetTag("shipment.estimated_days", shipment.EstimatedDays);

        _logger.LogInformation("Shipment scheduled: {ShipmentId} (class={ShippingClass} estimatedDays={EstimatedDays})",
            shipment.Id, command.ShippingClass, shipment.EstimatedDays);

        ShipmentsCounter.Add(1,
            new KeyValuePair<string, object?>("class", command.ShippingClass),
            new KeyValuePair<string, object?>("tier", tier));

        return shipment;
    }
}
